File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/validation/annotation/Username.java
package com.windea.demo.cloudcollect.core.validation.annotation;

import com.windea.demo.cloudcollect.core.validation.validator.UsernameValidator;

import javax.validation.Constraint;
import javax.validation.Payload;
import java.lang.annotation.*;

import static java.lang.annotation.ElementType.*;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

/**
 * 用户名的校验注解。
 */
@Documented
@Constraint(validatedBy = {UsernameValidator.class})
@Target({METHOD, FIELD, ANNOTATION_TYPE, CONSTRUCTOR, PARAMETER, TYPE_USE})
@Retention(RUNTIME)
public @interface Username {
	String message() default "Username is not correct.";

	Class<?>[] groups() default {};

	Class<? extends Payload>[] payload() default {};
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/validation/annotation/Username.kt
package com.windea.demo.cloudcollect.core.validation.annotation

import com.windea.demo.cloudcollect.core.validation.validator.*
import javax.validation.*
import kotlin.annotation.AnnotationTarget.*
import kotlin.reflect.*

/**用户名的校验注解。*/
@MustBeDocumented
@Constraint(validatedBy = [UsernameValidator::class])
@Target(FIELD, PROPERTY_GETTER)
annotation class Username(
	val message: String = "Username is not correct.",
	val groups: Array<KClass<*>> = [],
	val payload: Array<KClass<out Payload>> = []
)
